doschema/jsvalidation.py

```diff
@@ -0,0 +1,229 @@
+            if 'dependencies' in curr_schema:
+                self.traverse_dependencies(
+                    curr_schema,
+                    curr_field + ('dependencies', )
+                )
+        else:
+            for field_name, field_value in curr_schema.items():
+                # if a value is a dict recursively comes into it
+                if isinstance(field_value, dict):
+                    self.traverse(field_value, curr_field + (field_name, ))
+                elif field_name == 'type' or field_name == '$ref':
+                    self.process_type_field(field_value, curr_field)
+                elif field_name == 'items':
+                    if isinstance(field_value, list):
+                        self.process_array(field_value, curr_field)
+
+    def process_array(self, field_value, curr_field):
+        """Process each JSON object in the array according to ignore_index.
+
+        :param field_value: Type of the field.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        field_path = curr_field
+        for elem in field_value:
+            if self.ignore_index:
+                new_field_path = field_path
+            else:
+                new_field_path = field_path + (field_value.index(elem), )
+
+            self.traverse(elem, new_field_path)
+
+    def process_type_field(self, field_value, curr_field):
+        """Check if a name field with type "type" has already appeared.
+
+        If had appeared with different type,raise an exception.
+        Otherwise, add object to a dict of types.
+
+        :param field_value: Type of the field.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        if curr_field in self.fields_types_dict.keys():
+            if self.fields_types_dict[curr_field].field_type != field_value:
+                err_msg = "Type mismatch in schemas {0} and {1}."
+                old_schema = self.fields_types_dict[curr_field].schema_index
+                raise JSONSchemaCompatibilityException(
+                    old_schema,
+                    self.schema_name,
+                    err_msg.format(old_schema, self.schema_name)
+                )
+        else:
+            self.fields_types_dict[curr_field] = FieldToAdd(
+                schema_index=self.schema_name,
+                field_tuple=curr_field,
+                field_type=field_value
+            )
+
+    def check_special(self, curr_schema, curr_field):
+        """Check schema for allOf, anyOf and oneOf.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        keys = self.COLLECTION_KEYS.intersection(curr_schema)
+        for key in keys:
+            self.collect_special(
+                curr_schema[key],
+                curr_field
+            )
+
+    def collect_special(self, JSON_objects_list, curr_field):
```

Comment:
You are right that `JSON` should be written in upper case when it is alone or in a class name. But here the rule for variable naming is more important, thus lowercase.